pbgen: generate clients for other admin services

rockwotj opened the pull request:
- **serde/pb/rpc: add a context class**
- **admin/proxy: create a service for proxying internal RPC requests**
- **pbgen: pass in context to rpc methods**
- **pbgen: generate a client for each service**
- **admin: plumb in proxy support for admin RPC requests**

## Backports Required

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes

* none

rockwotj committed:
serde/pb/rpc: add a context class

To keep some context we'll want to pass around in RPC handlers.

rockwotj committed:
admin/proxy: create a service for proxying internal RPC requests

This will be used to ensure that our admin API acts "cluster wide",
and remove the hacky redirect logic that we have.

We pack and unpack the correct error type here so that we properly
propagate errors according to our RPC framework.

rockwotj committed:
pbgen: pass in context to rpc methods

In order to support proxying API calls to the correct leader, we want to
ensure that we don't have proxy loops or other kinds of bad behavior. In
order to prevent this, we'll add a context call to every RPC method that
will tell handers two things: if this node has already seen this API,
and it will also hold other information like if this is a proxied
request, which will be needed for scatter-gather RPCs that need to
collect information from all over the cluster.

Secondly, while the context is being passed in we also update the RPC
handler to have a simpler interface. This allows it to be reused in
non-HTTP scenarios (like when we proxy using our internal RPC
interface), and we can push more of the HTTP handling out of codegen and
instead to hand written code in the admin server.

rockwotj committed:
pbgen: generate a client for each service

rockwotj committed:
admin: plumb in proxy support for admin RPC requests

Hook up the proxy service to call into the admin server when it gets
requests, and also update the client to use the generated typed clients
for RPC requests to other nodes.

rockwotj commented:
I will push a test in the morning, but otherwise there is plenty of code to review in the meantime.

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--39469d2e-ae9f-4de7-b061-ac7232dbddf5-->
<!--TEST RESULTS-->
<!--7c23a257-cf29-4a67-bf24-f89ec2dfd140-->
<!--TEST RESULTS-->
<!--74e8aa7f-fdff-4687-b6b7-f9d02db613aa-->
<!--TEST RESULTS-->
<!--4122e554-3244-4980-95c2-e87b4b046e77-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/71195">build#71195</a></summary>


| test_class               | test_method                   | test_arguments                                                                                                                                 | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:-------------------------|:------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| CloudStorageScrubberTest | test_scrubber                 | {"cloud_storage_type": 2}                                                                                                                      | integration | https://buildkite.com/redpanda/redpanda/builds/71195#0198cea7-54c7-48b2-885e-4f208999a97e | FLAKY         | 19/21    | upstream reliability is '98.86039886039886'. current run reliability is '90.47619047619048'. drift is 8.38421 and the allowed drift is set to 50. The test should PASS |
| CrashReporterTest        | test_redpanda_crash_reporting | null                                                                                                                                           | integration | https://buildkite.com/redpanda/redpanda/builds/71195#0198cea7-54cd-4308-b5a6-a414668c1c07 | FLAKY         | 19/21    | upstream reliability is '100.0'. current run reliability is '90.47619047619048'. drift is 9.52381 and the allowed drift is set to 50. The test should PASS             |
| DataMigrationsApiTest    | test_transactions             | {"transfer_leadership": true}                                                                                                                  | integration | https://buildkite.com/redpanda/redpanda/builds/71195#0198cea7-54cc-45a3-a1e3-6cb63d3b9d12 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS              |
| RandomNodeOperationsTest | test_node_operations          | {"cloud_storage_type": 1, "compaction_mode": "chunked_sliding_window", "enable_failures": true, "mixed_versions": true, "with_iceberg": false} | integration | https://buildkite.com/redpanda/redpanda/builds/71195#0198cea4-d6e2-4d22-aa31-63c7fab8f684 | FLAKY         | 19/21    | upstream reliability is '100.0'. current run reliability is '90.47619047619048'. drift is 9.52381 and the allowed drift is set to 50. The test should PASS             |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/71240">build#71240</a></summary>


| test_class               | test_method              | test_arguments                                                                                                                         | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                               |
|:-------------------------|:-------------------------|:---------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:---------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| DatalakeE2ETests         | test_json_schema_unicode | {"catalog_type": "rest_hadoop", "cloud_storage_type": 1, "query_engine": "trino"}                                                      | integration | https://buildkite.com/redpanda/redpanda/builds/71240#0198d22d-1390-47ea-83f2-9631eec29382 | FLAKY         | 20/21    | upstream reliability is '98.37209302325581'. current run reliability is '95.23809523809523'. drift is 3.134 and the allowed drift is set to 50. The test should PASS |
| RandomNodeOperationsTest | test_node_operations     | {"cloud_storage_type": 2, "compaction_mode": "adjacent_merge", "enable_failures": true, "mixed_versions": true, "with_iceberg": false} | integration | https://buildkite.com/redpanda/redpanda/builds/71240#0198d22d-138d-43dd-9f25-59d6d22ecece | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS            |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/71279">build#71279</a></summary>


| test_class       | test_method   | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                 |
|:-----------------|:--------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:---------------------------------------|
| AdminApiAuthTest | test_admin_v2 | null             | integration | https://buildkite.com/redpanda/redpanda/builds/71279#0198d392-0319-4145-92f0-a82d69a37d3b | FAIL          | 0/21     | The test has failed across all retries |
| AdminApiAuthTest | test_admin_v2 | null             | integration | https://buildkite.com/redpanda/redpanda/builds/71279#0198d398-b9b6-4358-a96d-d665aacebe5e | FAIL          | 0/1      |                                        |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/71298">build#71298</a></summary>


| test_class                       | test_method              | test_arguments                                                                                                                         | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:---------------------------------|:-------------------------|:---------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| AutomaticLeadershipBalancingTest | test_automatic_rebalance | null                                                                                                                                   | integration | https://buildkite.com/redpanda/redpanda/builds/71298#0198d46f-5c93-4924-98aa-a7f5027a6064 | FLAKY         | 20/21    | upstream reliability is '99.71751412429379'. current run reliability is '95.23809523809523'. drift is 4.47942 and the allowed drift is set to 50. The test should PASS |
| RandomNodeOperationsTest         | test_node_operations     | {"cloud_storage_type": 1, "compaction_mode": "adjacent_merge", "enable_failures": true, "mixed_versions": true, "with_iceberg": false} | integration | https://buildkite.com/redpanda/redpanda/builds/71298#0198d46f-5c94-4c88-b93c-0c866e187a23 | FLAKY         | 19/21    | upstream reliability is '100.0'. current run reliability is '90.47619047619048'. drift is 9.52381 and the allowed drift is set to 50. The test should PASS             |
| RandomNodeOperationsTest         | test_node_operations     | {"cloud_storage_type": 2, "compaction_mode": "sliding_window", "enable_failures": true, "mixed_versions": true, "with_iceberg": false} | integration | https://buildkite.com/redpanda/redpanda/builds/71298#0198d46f-5c95-4ac0-8f10-ca519e43bd0b | FLAKY         | 19/21    | upstream reliability is '99.49238578680203'. current run reliability is '90.47619047619048'. drift is 9.0162 and the allowed drift is set to 50. The test should PASS  |
</details>
<!--/body-->

rockwotj commented:
Force push: fix up deps from header files

michael-redpanda commented on src/v/redpanda/admin/proxy/client.h:
```diff
@@ -0,0 +1,46 @@
+namespace admin::proxy {
+
+// This is a generic client to proxy any ConnectRPC request to another node in
+// the clister. It can be used to forward mutations to the correct leader, or it
```
nit: s/clister/cluster

michael-redpanda commented on src/v/redpanda/admin/proxy/client.cc:
```diff
@@ -0,0 +1,149 @@
+          target,
+          ctx.proxied_nodes());
+        throw serde::pb::rpc::unavailable_exception("proxy loop detected");
+    }
```
question: this could naturally occur due to leadership transfers right?  So the client would just have to resend the request?

Would `508` be a more appropriate response (https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Status/508)?

rockwotj replied:
> question: this could naturally occur due to leadership transfers right? So the client would just have to resend the request?

Yes.

> Would 508 be a more appropriate response (https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Status/508)?

Not available in our RPC framework: https://connectrpc.com/docs/protocol#error-codes

If we want to embed more information to mention that a request is retryable, then we can do that with `details`: https://connectrpc.com/docs/protocol#error-end-stream

michael-redpanda replied:
> If we want to embed more information to mention that a request is retryable, then we can do that with details

Do we want to do that now or should we just instruct consumers of this API to always do a retry?

rockwotj replied:
Let me add `details` to errors in a followup.

michael-redpanda commented on src/v/redpanda/admin/proxy/types.h:
```diff
@@ -0,0 +1,70 @@
+
+namespace admin::proxy {
+
+struct proxy_request
+  : serde::
+      envelope<proxy_request, serde::version<0>, serde::compat_version<0>> {
+    ss::sstring service; // For example: "redpanda.core.admin.Service"
+    ss::sstring method;  // For example: "MyRPCMethod"
+    iobuf payload;       // The serialized request protobuf
+    // Nodes that have already proxied this request.
+    std::vector<model::node_id> via;
+
+    auto serde_fields() { return std::tie(service, method, payload, via); }
+};
```
nit: do we care about having a `fmt::formatter` for this and `proxy_response`?

rockwotj replied:
Not at the moment, but easy enough to add with `format_to`

michael-redpanda commented on src/v/redpanda/admin/proxy/client.cc:
```diff
@@ -0,0 +1,149 @@
+} // namespace
+
+ss::future<iobuf> client::send(
+  model::node_id target, serde::pb::rpc::context ctx, iobuf payload) noexcept {
```
question: do we need to check that the target != _self?  Would that be an error?

rockwotj replied:
We could check for that, I think the RPC subsystem will throw an error that there is no connection to `self`.

michael-redpanda replied:
I guess would sending it to itself result in a 503 unavailable error because the node already exists in the `via` list?

rockwotj replied:
yeah it will return a 503, but for different reasons :)

michael-redpanda commented on src/v/redpanda/admin/server.cc:
```diff
@@ -350,69 +353,151 @@ namespace {
+            iobuf reply_payload = co_await _descriptor.handler(
+              std::move(ctx), std::move(request_payload));
+            rep->write_body(
+              "bin",
```
question: should this always be `bin`?  should it be `bin` or `json` depending on the `ctx.content_type`?

rockwotj replied:
Yeah, I can add back the comment, but seastar doesn't have a content type for `application/proto` so we just write a dummy value then overwrite it in the next line

michael-redpanda commented on src/v/redpanda/admin/proxy/client.h:
```diff
@@ -28,6 +27,17 @@ class client {
       : _self(self)
       , _conn_cache(conn_cache) {}
 
+    // Create an RPC client for the given node, which can be used to
+    // proxy requests to that node.
+    template<typename ProtobufServiceClient>
+    ProtobufServiceClient make_client_for_node(model::node_id node) {
+        return ProtobufServiceClient(
+          [this, node](serde::pb::rpc::context ctx, iobuf payload) {
+              return send(node, std::move(ctx), std::move(payload));
+          });
+    }
```
question: this used anywhere?

rockwotj replied:
not yet, but this is how you're supposed to make the RPC clients that we codegen now.

Let me make the existing admin endpoints for build info and listing routes to be cluster-wide.

rockwotj replied:
Okay added some tests (even for error cases).

michael-redpanda reviewed (commented):
very nice

rockwotj committed:
proto/admin: update admin protos to have cluster-wide scope

Mostly in order to demonstrate usage of this proxy client, show two
different implementations of request styles I anticipate with the new
admin APIs being cluster scoped.

1. There is a clear node that should service the request (in this
   manner by specifying it in the request). We need to either serve the
   request or proxy to the right node for it.

2. We need some aggregated view of the cluster (in this case the build
   info for each broker), so we make requests to all other brokers and
   aggregate them in the handler.

rockwotj commented on src/v/redpanda/admin/server.cc:
```diff
@@ -350,73 +353,176 @@ namespace {
+                    std::move(writer), payload.share(0, payload.size_bytes()));
+              });
+            rep->set_mime_type(
+              ctx.content_type == serde::pb::rpc::content_type::proto
```
right that there is an issue here, wrong what the issue actually is

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/71279-->
<!--983a55fd-a6eb-4311-bb2d-cf78f2da9c35-->
<!--title-->
## Retry command for [Build#71279](https://buildkite.com/redpanda/redpanda/builds/71279)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/admin_api_auth_test.py::AdminApiAuthTest.test_admin_v2
```
<!--/body-->

oleiman reviewed (approved):
cool stuff. i haven't been keeping such close tabs on the new admin code, but I think this all lgtm. 

one request: is there any way we could have the gen'ed code marked as such? I walk away from these diffs with the impression that I both reviewed generated code and missed hand-generated code.

rockwotj commented:
> cool stuff. i haven't been keeping such close tabs on the new admin code, but I think this all lgtm. 
> 
> 
> 
> one request: is there any way we could have the gen'ed code marked as such? I walk away from these diffs with the impression that I both reviewed generated code and missed hand-generated code. 

Ah yes, we can: https://docs.github.com/en/repositories/working-with-files/managing-files/customizing-how-changed-files-appear-on-github

Will do inna followup

rockwotj commented:
> > cool stuff. i haven't been keeping such close tabs on the new admin code, but I think this all lgtm.
> > one request: is there any way we could have the gen'ed code marked as such? I walk away from these diffs with the impression that I both reviewed generated code and missed hand-generated code.
> 
> Ah yes, we can: https://docs.github.com/en/repositories/working-with-files/managing-files/customizing-how-changed-files-appear-on-github
> 
> Will do inna followup

Hopefully this will do the trick: https://github.com/redpanda-data/redpanda/pull/27375

michael-redpanda commented:
damn you beat me :(

rockwotj commented:
Should hopefully be a trivial change to add the context parameter
